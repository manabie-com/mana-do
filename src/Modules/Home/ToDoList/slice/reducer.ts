import {Todo, TodoStatus} from '../../../../models/todo';
import {
  AppActions,
  CREATE_TODO,
  DELETE_ALL_TODOS,
  DELETE_TODO,
  INIT_TODO,
  INIT_TODO_DONE,
  TOGGLE_ALL_TODOS,
  UPDATE_IS_IMPORTANT,
  UPDATE_STATUS_SHOWING,
  UPDATE_TODO,
  UPDATE_TODO_STATUS
} from './actions';
import {MenuActive} from "../../../../components/molecules/Toolbar";
import {MAP_SHOWING_TO_ACTIVE_MENU} from "./constants";

export interface AppState {
  todos: Array<Todo>,
  isLoading: boolean,
  isInit: boolean,
  showing: TodoStatus,
  menuActive: MenuActive,
}

export const initialState: AppState = {
  todos: [],
  isLoading: true,
  isInit: false,
  showing: TodoStatus.ALL,
  menuActive: MenuActive.ALL,
}

function reducer(state: AppState, action: AppActions): AppState {
  switch (action.type) {
    case INIT_TODO:
      return {
        ...state,
        isInit: false,
        isLoading: true
      };
    case INIT_TODO_DONE:
      return{
        ...state,
        todos: action.payload,
        isInit: true,
        isLoading: false
      }
    case CREATE_TODO:
      return {
        ...state,
        todos: [...state.todos, action.payload]
      };

    case UPDATE_TODO_STATUS:
      const index2 = state.todos.findIndex((todo) => todo.id === action.payload.todoId);
      state.todos[index2].status = action.payload.checked ? TodoStatus.COMPLETED : TodoStatus.ACTIVE;
      return {
        ...state,
        todos: state.todos
      }

    case TOGGLE_ALL_TODOS:
      const tempTodos = state.todos.map((e)=>{
        return {
          ...e,
          status: action.payload ? TodoStatus.COMPLETED : TodoStatus.ACTIVE
        }
      })

      return {
        ...state,
        todos: tempTodos
      }

    case DELETE_TODO:
      const index1 = state.todos.findIndex((todo) => todo.id === action.payload);
      const dataTmp = [...state.todos];
      dataTmp.splice(index1, 1);

      return {
        ...state,
        todos: dataTmp
      }
    case DELETE_ALL_TODOS:
      return {
        ...state,
        todos: [],
        menuActive: MenuActive.CLEAR
      }

    case UPDATE_STATUS_SHOWING:
      return {
        ...state,
        showing: action.payload,
        menuActive: MAP_SHOWING_TO_ACTIVE_MENU[action.payload] as MenuActive
      }

    case UPDATE_IS_IMPORTANT:
      const {todoId, isImportant} = action.payload;
      const indexItemTodo = state.todos.findIndex((todo) => todo.id === todoId);
      const dataClone = [...state.todos];
      if(indexItemTodo !== -1){
        dataClone[indexItemTodo].is_important = isImportant
      }

      return {
        ...state,
        todos: dataClone
      }
    case UPDATE_TODO:
      const indexItem = state.todos.findIndex(todo => todo.id === action.payload.todoId);

      const todoClone = [...state.todos];

      if(indexItem !== -1){
        todoClone[indexItem].content = action.payload.content
      }

      return {
        ...state,
        todos: [...todoClone]
      }
    default:
      return state;
  }
}

export default reducer;