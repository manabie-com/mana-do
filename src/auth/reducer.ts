import {
  Action,
  REQUEST,
  REQUEST_FAIL,
  AUTHENTICATE_SUCCESS,
  LOGOUT_SUCCESS,
} from './actions';

export interface AuthProviderState {
  token: string;
  isLoading: boolean;
  error: string;
}

export function reducer(
  state: AuthProviderState,
  action: Action
): AuthProviderState {
  switch (action.type) {
    case REQUEST:
      return { ...state, isLoading: true };
    case REQUEST_FAIL:
      const error = action.error;
      return { ...state, isLoading: false, error };
    case AUTHENTICATE_SUCCESS:
      const token = action.token;
      return { ...state, isLoading: false, error: '', token };
    case LOGOUT_SUCCESS:
      return { ...state, token: '' };
    default:
      return state;
  }
}
