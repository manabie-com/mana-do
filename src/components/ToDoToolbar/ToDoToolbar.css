.Todo__toolbar {
	display: flex;
	justify-content: space-between;
	margin-top: 24px;
}

.Todo__tabs {
	display: flex;
	justify-content: center;
}

.Todo__tabs > *:not(:last-child) {
	margin-right: 8px;
}

.Action__btn.active{
	background-color: aquamarine;
}