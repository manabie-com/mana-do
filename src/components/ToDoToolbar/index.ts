export * from './ToDoToolbar';
export {default} from './ToDoToolbar';