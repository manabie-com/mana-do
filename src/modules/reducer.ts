import Actions, { types } from "./actions";

export interface GlobalState {
  isLogined: boolean;
}

const initialState: GlobalState = {
  isLogined: localStorage.getItem('token') !== "",
};

const globalReducer = (state = initialState, action: Actions): GlobalState => {
  const newState = { ...state };

  switch (action.type) {
    case types.LOGIN:
      newState.isLogined = true;
      localStorage.setItem("token", action.payload);
      return newState;
    case types.LOGOUT:
      newState.isLogined = false;
      localStorage.removeItem("token");
      return newState;
    default:
      return state;
  }
};

export default globalReducer;
