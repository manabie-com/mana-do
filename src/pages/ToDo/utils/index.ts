import { Todo, TodoStatus } from '../../../models/todo';
const created_date = new Date().toISOString();
export const defaultToDos = {
  content: 'Content',
  created_date,
  status: TodoStatus.ACTIVE,
  id: 'testId',
  user_id: 'firstUser',
} as Todo;

export function spawnToDos(count: number): Todo[] {
  return new Array(count).fill(defaultToDos).map(todo => {
    todo.id = `testId-${Math.random()}`;
    return todo;
  });
}
