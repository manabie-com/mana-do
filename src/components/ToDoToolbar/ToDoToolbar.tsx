import React from 'react';
import { EnhanceTodoStatus, Todo, TodoStatus } from '../../models';
import { toCapitalize } from '../../utils';
import './ToDoToolbar.css'

export interface ToDoToolbarProps {
    todos: Todo[],
    showing: EnhanceTodoStatus,
    remainTodos: number,
    onToggleAllTodo(e: React.ChangeEvent<HTMLInputElement>): void,
    setShowing(todoStatus: EnhanceTodoStatus): void,
    onDeleteAllTodo(): void
}

function ToDoToolbar({todos, showing, remainTodos, onToggleAllTodo, setShowing, onDeleteAllTodo}: ToDoToolbarProps) {
    const filterButtons: EnhanceTodoStatus[] = ["ALL", TodoStatus.ACTIVE, TodoStatus.COMPLETED];

    return (
        <div className="Todo__toolbar">
                {todos.length > 0 ?
                    ( 
                    <>
                        <input
                            id="toggle-all"
                            type="checkbox"
                            className='toggle'
                            onChange={onToggleAllTodo}
                        />
                        <label className='checkbox-custom' htmlFor="toggle-all"></label>
                    </>
                    ): <div className='Todo__empty-toggle'/>
                }
                <div className="Todo__tabs">
                    {
                        filterButtons.map((button: EnhanceTodoStatus, index) => {
                            return (
                                <button 
                                    key={index} 
                                    className={"Action__btn" + (showing === button ? " active" : "")} 
                                    onClick={()=>setShowing(button)}>
                                    {toCapitalize(button)}
                                </button>
                            );
                        })
                    }
                    <button 
                        className="Action__btn Todo__btn-delete-all" 
                        disabled={todos.length <= 0} 
                        onClick={onDeleteAllTodo}>
                        Clear all todos
                    </button>
                </div>
                <span className='Todo__remain-item'>{remainTodos} item{remainTodos > 0 ? 's' : ''} left</span>
            </div>
    )
}

export default ToDoToolbar;