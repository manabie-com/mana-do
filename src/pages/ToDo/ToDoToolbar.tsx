import React from 'react';
import { AppContext } from '../../context/toDoContext';
import { TodoStatus } from '../../models/todo';
import {
  deleteAllTodos,
  filterTodos,
  toggleAllTodos,
} from '../../store/actions';
import {
  getAllToDosStatus,
  getViewOptions,
  saveSelectionOptions,
} from '../../utils/localStorage';
import { playCheckedEffect } from '../../utils/mojs';

export type EnhanceTodoStatus = TodoStatus | 'ALL';

export default function ToDoToolbar() {
  const { state, dispatch } = React.useContext(AppContext);
  const toggleAllTodoRef = React.useRef<any>(null);

  const onToggleAllTodo = (e: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(toggleAllTodos(e.target.checked));
  };

  const onDeleteAllTodo = () => {
    dispatch(deleteAllTodos());
    toggleAllTodoRef.current.checked = false;
  };

  const checkedEffect = (e: React.MouseEvent<HTMLInputElement>) => {
    const { checked } = e.currentTarget;
    const mousPosition = { x: e.clientX, y: e.clientY };
    if (checked) {
      playCheckedEffect(mousPosition);
    }
  };

  const onFilterTodos = (e: React.ChangeEvent<HTMLSelectElement>) => {
    dispatch(filterTodos(e.target.value as EnhanceTodoStatus));
    saveSelectionOptions(e.target.value as EnhanceTodoStatus);
  };

  const defaultChecked =
    getAllToDosStatus(state.todos) === TodoStatus.COMPLETED;
  return (
    <div className='Todo__toolbar'>
      {state.todos.length > 0 ? (
        <div className='Toolbar__toggle_all'>
          <input
            ref={toggleAllTodoRef}
            id='toggle-all'
            aria-label='toggle-all-todos'
            type='checkbox'
            defaultChecked={defaultChecked}
            onChange={onToggleAllTodo}
            onClick={checkedEffect}
          />
        </div>
      ) : (
        <div />
      )}
      <div className='Todo__tabs'>
        <label htmlFor='filter'>View</label>
        <select
          defaultValue={getViewOptions()}
          name='filter'
          id='filter'
          onChange={onFilterTodos}
        >
          <option value='ALL'>All</option>
          <option value='ACTIVE'>Active</option>
          <option value='COMPLETED'>Completed</option>
        </select>
      </div>
      <button
        disabled={Boolean(state.todos.length === 0)}
        className='Action__btn'
        onClick={onDeleteAllTodo}
        arial-label='delete-all-todos'
      >
        Clear all todos
      </button>
    </div>
  );
}
