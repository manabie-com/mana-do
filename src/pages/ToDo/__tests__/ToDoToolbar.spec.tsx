import React from 'react';
import ToDoToolbar from '../ToDoToolbar';
import { render } from '@testing-library/react';
import { AppContext } from '../../../context/toDoContext';
import { spawnToDos } from '../utils';
import { AppState } from '../../../store/reducer';

describe('ToDoToolbar', () => {
  it('Should match snapshot', () => {
    const { container } = render(<ToDoToolbar />);
    expect(container).toMatchSnapshot();
  });
});

describe('Render ToDoToolbar', () => {
  it('Toggle all checkbox should not checked', async () => {
    const todos = spawnToDos(1);
    const state = { todos, filter: 'ALL' } as AppState;
    const { findByLabelText } = render(
      <AppContext.Provider value={{ state, dispatch: jest.fn() }}>
        <ToDoToolbar />
      </AppContext.Provider>,
    );
    expect(await findByLabelText(/toggle-all-todos/i)).not.toBeChecked();
  });

  it('Toggle all checkbox should checked', async () => {
    const todos = spawnToDos(1);
    todos[0].status = 'COMPLETED';
    const state = { todos, filter: 'ALL' } as AppState;
    const { findByLabelText } = render(
      <AppContext.Provider value={{ state, dispatch: jest.fn() }}>
        <ToDoToolbar />
      </AppContext.Provider>,
    );
    expect(await findByLabelText(/toggle-all-todos/i)).toBeChecked();
  });

  it('Should render select element', () => {
    const { getByLabelText } = render(<ToDoToolbar />);
    expect(getByLabelText(/view/i)).toBeInTheDocument();
  });

  it('Should render Clear all todos button', () => {
    const { getByText } = render(<ToDoToolbar />);
    expect(getByText(/clear all todos/i)).toBeInTheDocument();
  });
});
