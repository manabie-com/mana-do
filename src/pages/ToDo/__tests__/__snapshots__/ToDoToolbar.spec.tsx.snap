// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoToolbar Should match snapshot 1`] = `
<div>
  <div
    class="Todo__toolbar"
  >
    <div />
    <div
      class="Todo__tabs"
    >
      <label
        for="filter"
      >
        View
      </label>
      <select
        id="filter"
        name="filter"
      >
        <option
          selected=""
          value="ALL"
        >
          All
        </option>
        <option
          value="ACTIVE"
        >
          Active
        </option>
        <option
          value="COMPLETED"
        >
          Completed
        </option>
      </select>
    </div>
    <button
      arial-label="delete-all-todos"
      class="Action__btn"
      disabled=""
    >
      Clear all todos
    </button>
  </div>
</div>
`;
